manager/scheduler/filter.go

```diff
@@ -359,3 +359,34 @@ func (f *HostPortFilter) Explain(nodes int) string {
 	}
 	return fmt.Sprintf("host-mode port already in use on %d nodes", nodes)
 }
+
+// MaxReplicasFilter selects only nodes that does not exceed max replicas per node.
+type MaxReplicasFilter struct {
+	t *api.Task
+}
+
+// SetTask returns true when max replicas per node filter > 0 for a given task.
+func (f *MaxReplicasFilter) SetTask(t *api.Task) bool {
+	if t.Spec.Placement != nil {
+		if t.Spec.Placement.Maxreplicas > 0 {
+			f.t = t
+			return true
+		}
+	}
+
+	return false
+}
+
+// Check returns true if there is free slots for task in a given node.
+func (f *MaxReplicasFilter) Check(n *NodeInfo) bool {
+	if uint64(n.ActiveTasksCountByService[f.t.ServiceID]) < f.t.Spec.Placement.Maxreplicas {
```

Comment:
is this concurrency safe? if so, can you explain why it's concurrency safe in a comment?

Reply:
As far I understand swarmkit code then scheduler will process only one task at time?
and code logic on https://github.com/docker/swarmkit/blob/master/manager/scheduler/nodeinfo.go#L114 updates ActiveTasksCountByService counter immediately when task is scheduled to host without waiting that it actually starts. That should make it concurrency safe.

But to be honest I'm not sure how to describe it nicely.